sys/hashes/sha3.c

```diff
@@ -0,0 +1,463 @@
+   LITTLE_ENDIAN symbol in that case.
+
+   For a more complete set of implementations, please refer to
+   the Keccak Code Package at https://github.com/gvanas/KeccakCodePackage
+
+   For more information, please refer to:
+ * [Keccak Reference] http://keccak.noekeon.org/Keccak-reference-3.0.pdf
+ * [Keccak Specifications Summary] http://keccak.noekeon.org/specs_summary.html
+
+   This file uses UTF-8 encoding, as some comments use Greek letters.
+   ================================================================
+ */
+
+#include <hashes/sha3.h>
+
+/**
+ * Function to compute the Keccak[r, c] sponge function over a given input.
+ * @param  rate            The value of the rate r.
+ * @param  capacity        The value of the capacity c.
+ * @param  input           Pointer to the input message.
+ * @param  inputByteLen    The number of input bytes provided in the input message.
+ * @param  delimitedSuffix Bits that will be automatically appended to the end
+ *                         of the input message, as in domain separation.
+ *                         This is a byte containing from 0 to 7 bits
+ *                         These <i>n</i> bits must be in the least significant bit positions
+ *                         and must be delimited with a bit 1 at position <i>n</i>
+ *                         (counting from 0=LSB to 7=MSB) and followed by bits 0
+ *                         from position <i>n</i>+1 to position 7.
+ *                         Some examples:
+ *                             - If no bits are to be appended, then @a delimitedSuffix must be 0x01.
+ *                             - If the 2-bit sequence 0,1 is to be appended (as for SHA3-*), @a delimitedSuffix must be 0x06.
+ *                             - If the 4-bit sequence 1,1,1,1 is to be appended (as for SHAKE*), @a delimitedSuffix must be 0x1F.
+ *                             - If the 7-bit sequence 1,1,0,1,0,0,0 is to be absorbed, @a delimitedSuffix must be 0x8B.
+ * @param  output          Pointer to the buffer where to store the output.
+ * @param  outputByteLen   The number of output bytes desired.
+ * @pre    One must have r+c=1600 and the rate a multiple of 8 bits in this implementation.
+ */
+void Keccak(unsigned int rate, unsigned int capacity, const unsigned char *input, unsigned long long int inputByteLen, unsigned char delimitedSuffix, unsigned char *output, unsigned long long int outputByteLen);
+
+/**
+ *  Function to compute SHAKE128 on the input message with any output length.
+ */
+void FIPS202_SHAKE128(const unsigned char *input, unsigned int inputByteLen, unsigned char *output, int outputByteLen)
+{
+    Keccak(1344, 256, input, inputByteLen, 0x1F, output, outputByteLen);
+}
+
+/**
+ *  Function to compute SHAKE256 on the input message with any output length.
+ */
+void FIPS202_SHAKE256(const unsigned char *input, unsigned int inputByteLen, unsigned char *output, int outputByteLen)
```

Comment:
Same comments apply as for [this](https://github.com/RIOT-OS/RIOT/pull/7881/files#diff-2b0dd83345694afd13187035fff86039R85) apply here here